openmdao/core/system.py

```diff
@@ -4648,26 +4636,35 @@ def _get_vars_exec_order(self, inputs=False, outputs=False, variables=None):
         real_vars = self._var_allprocs_abs2meta
         disc_vars = self._var_allprocs_discrete
 
-        in_or_out = []
+        # variable order: real inputs, real outputs, discrete inputs, discrete outputs
+        var_dicts = []
         if inputs:
-            in_or_out.append('input')
+            var_dicts.append(real_vars['input'])
         if outputs:
-            in_or_out.append('output')
+            var_dicts.append(real_vars['output'])
+        if inputs:
+            var_dicts.append(disc_vars['input'])
+        if outputs:
+            var_dicts.append(disc_vars['output'])
 
+        # For components with no children, self._subsystems_allprocs is empty.
         if self._subsystems_allprocs:
-            for subsys, _ in self._subsystems_allprocs.values():
+            if local:
+                from openmdao.core.component import Component
+                it = self.system_iter(recurse=True, typ=Component)
+            else:
+                it = iter(subsys for subsys, _ in self._subsystems_allprocs.values())
+
+            for subsys in it:
                 prefix = subsys.pathname + '.'
-                for io in in_or_out:
-                    for var_name in chain(real_vars[io], disc_vars[io]):
-                        if variables is None or var_name in variables:
-                            if var_name.startswith(prefix):
-                                var_list.append(var_name)
-        else:
-            # For components with no children, self._subsystems_allprocs is empty.
-            for io in in_or_out:
-                for var_name in chain(real_vars[io], disc_vars[io]):
+                for var_name in chain(*var_dicts):
                     if not variables or var_name in variables:
-                        var_list.append(var_name)
+                        if var_name.startswith(prefix):
+                            var_list.append(var_name)
+        else:
+            for var_name in chain(*var_dicts):
+                if not variables or var_name in variables:
```

Comment:
'not variables' should really be 'variables is None' here since passing an empty 'variables' shouldn't match anything.